### PR TITLE
compact: Additional test for issue prometheus#3943

### PR DESCRIPTION
**Pull request by Bplotka**
Really in debugging form just to figure out the problem now. [Not prepared to be merged]

TL;DR: Found bug - got repro of case when compaction is merging `36h` + `6h` block into `54h` with `12h` block gap DESPITE the fact 12h block was there! This is because we ignore last block in compaction - and it results of 2 block being compacted and USED without one 12h laying around and not being used. Nothing got's deleted. 

Fix in progress in separate PR

I tried to mimick the compaction flow from the logs provided by @hgranillo in https://github.com/prometheus/prometheus/issues/3943 

Further weirdness caused by the bug:
`level=info ts=2018-03-09T15:05:31.944625635Z caller=compact.go:394 component=tsdb msg="compact blocks" count=3 mint=1520402400000 maxt=1520596800000`
and couple of logs after:
`level=info ts=2018-03-09T17:02:51.604621507Z caller=compact.go:394 component=tsdb msg="compact blocks" count=2 mint=1520402400000 maxt=1520575200000`

How compaction can REDUCE block?

### CODE REVIEW

**Comment by Bplotka** [comment]
Ok so to 10th iteration - so far so good

**Comment by fabxc** [comment]
Thanks. It's quite hard to say if it's not entirely known what blocks existed at what point.

But the 12h thing can happen if the 18h slice (modulo aligned by unix seconds) doesn't have any blocks for the first 6h, e.g. after startup. Then we'll just create the 12h block for the last 2/3 of the slice.

**Comment by Bplotka** [comment]
OK so moving forward, until I can repro mentioned logs

**Comment by Bplotka** [comment]
Closing in favor of https://github.com/prometheus/tsdb/pull/299
